### PR TITLE
Batch relayer: Silo

### PR DESCRIPTION
**Pull request by baileyspraggins**
# Description

Including batch relayer and tests for Silo protocol.

## Type of change

- [ ] Bug fix 
- [x] New feature 
- [ ] Breaking change 
- [ ] Dependency changes
- [ ] Code refactor / cleanup
- [ ] Documentation or wording changes
- [ ] Other

## Checklist:

- [x] The diff is legible and has no extraneous changes
- [x] Complex code has been commented, including external interfaces
- [x] Tests are included for all code paths
- [x] The base branch is either `master`, or there's a description of how to merge

### CODE REVIEW

File: pkg/interfaces/contracts/standalone-utils/ISilo.sol [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        // Token that represents a share in totalDeposits of Silo
+        IShareToken collateralToken;
+        // Token that represents a share in collateralOnlyDeposits of Silo
+        IShareToken collateralOnlyToken;
+        // Token that represents a share in totalBorrowAmount of Silo
+        IShareToken debtToken;
+        // COLLATERAL: Amount of asset token that has been deposited to Silo with interest earned by depositors.
+        // It also includes token amount that has been borrowed.
+        uint256 totalDeposits;
+        // COLLATERAL ONLY: Amount of asset token that has been deposited to Silo that can be ONLY used
```
Comment (EndymionJkb) [inline pkg/interfaces/contracts/standalone-utils/ISilo.sol:32]:
```suggestion
        // COLLATERAL ONLY: Amount of asset token that has been deposited to Silo that can ONLY be used
```

File: pkg/standalone-utils/contracts/relayer/SiloWrapping.sol [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        // to be sourced from outside the relayer, we must first them pull them here.
+        if (sender != address(this)) {
+            require(sender == msg.sender, "Incorrect sender");
+            _pullToken(sender, wrappedToken, amount);
+        }
+
+        // No approval is needed here, as the shareTokens are burned directly from the relayer's account.
+        // Setting the amount to uint256(-1) informs Silo that we'd like to redeem all the relayer's shares.
+        // Ignore the return value which cannot be trusted. It does not include any fees assessed.
+        silo.withdraw(address(underlyingToken), type(uint256).max, false);
```
Comment (EndymionJkb) [inline pkg/standalone-utils/contracts/relayer/SiloWrapping.sol:94]: Comment says (-1); code is "max". Could reconcile by using a constant:
uint256 private constant MAX_UINT256 = type(uint256).max;

File: pkg/standalone-utils/contracts/test/MockShareToken.sol [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+contract MockShareToken is TestToken, IShareToken {
+    ISilo private immutable _silo;
+    address private immutable _asset;
+
+    /**
+     * @dev Token is always deployed for specific Silo and asset
+     * @param name token name
+     * @param symbol token symbol
+     * @param silo Silo address for which tokens was deployed
+     * @param asset asset for which this tokens was deployed
```
Comment (EndymionJkb) [inline pkg/standalone-utils/contracts/test/MockShareToken.sol:31]:
```suggestion
     * @param silo Silo address at which tokens were deployed
     * @param asset Asset for which these tokens were deployed
```

File: pkg/standalone-utils/contracts/test/MockSilo.sol [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    }
+
+    function withdraw(
+        address _asset,
+        uint256 _amount,
+        bool /*_collateralOnly*/
+    ) external override returns (uint256 withdrawnAmount, uint256 withdrawnShare) {
+        address shareTokenAddress = address(_assetStorage[_asset].collateralToken);
+        uint256 burnedShare;
+        if (_amount == type(uint256).max) {
```
Comment (EndymionJkb) [inline pkg/standalone-utils/contracts/test/MockSilo.sol:89]: Like the others, could be a constant, either MAX_UINT256 with a comment about what that means, or something self-documenting like ENTIRE_BALANCE_SENTINEL_VALUE

Comment (baileyspraggins) [reply pkg/standalone-utils/contracts/test/MockSilo.sol:89]: Whats the benefit to having type(uint256).max being set as a constant? Wouldn't it cost more to make it a constant and read it during comparison operations vs leaving it as is and adding a comment explaining the logic.

Comment (EndymionJkb) [reply pkg/standalone-utils/contracts/test/MockSilo.sol:89]: There's no cost to read a constant value (they're interpreted at compile-time and baked into the bytecode); it's just for documentation. If you only use it once I guess it doesn't matter if it's a constant vs a comment. A descriptive constant would be self-documenting (i.e., you wouldn't need to read the comment to understand it), but I don't feel strongly.

**Review by EndymionJkb — APPROVED** [review]

**Pull request merged**